proto/substrait/algebra.proto

```diff
@@ -342,6 +342,34 @@ message ExchangeRel {
   }
 }
 
+// Duplicates records, possibly switching output expressions between each duplicate.
+// Default output is all of the fields declared followed by one int64 field that contains the
+// duplicate_id which is a zero-index ordinal of which duplicate of the original record this
+// corresponds to.
+message ExpandRel {
+  RelCommon common = 1;
+  Rel input = 2;
+  repeated ExpandField fields = 4;
+
+  message ExpandField {
+    oneof field_type {
+      // Field that switches output based on which duplicate_id we're outputting
+      SwitchingField switching_field = 2;
+
+      // Field that outputs the same value no matter which duplicate_id we're on.
+      Expression consistent_field = 3;
+    }
+  }
+
+  message SwitchingField {
+    // Array that contains an expression to output per duplicate_id
+    // each `switching_field` must have the same number of expressions
+    // all expressions within a switching field must be the same type class but can differ in nullability.
+    // this column will be nullable if any of the expressions are nullable.
+    repeated Expression duplicates = 1;
```

Comment:
Is there any value in having a repeated inside a repeated?  Seems like this should just be a single item (to be consistent with consistent_field above).

Reply:
The outer repeated (`fields`) is columns and the inner repeated (`duplicates`) is rows.

So if our goal is to take input:

| X | Y | Z |
| --- | --- | --- |
| 1 | 2 | 3 |

and generate:

| X | Y | Z |
| --- | --- | --- |
| 1 | 2 | 3 |
| 1 | NULL | 3 |
| NULL | NULL | 3 |

Then we need something like...

```
fields: [
  { duplicates: [ field(x), field(x), NULL ] },
  { duplicates: [ field(y), NULL, NULL ] }
]
```

Reply:
Thanks for the clarification.  I've sent #546 to expand the documentation.